Fairness metric

msmuskan opened the pull request.

msmuskan committed:
Create README.md

msmuskan committed:
Merge remote-tracking branch 'upstream/etm-fairness' into etm-fairness

msmuskan committed:
Merge remote-tracking branch 'upstream/etm-fairness' into etm-fairness

msmuskan committed:
avg costs; hashmap of objects

msmuskan committed:
Runtime settings

Increase permsize,
disable assertions

msmuskan committed:
Update average cost calculation

Handle divide by zero or null.

Handle variable (group) is null, by not printing the value.

msmuskan committed:
Correct and simplify fairness calculation

- Add  1 to avgA, avgB (very important change; resulted in a perfect process model, when tested on simple Python artificial log)

- Use double instead of AtomicLong variables (to simplify the code, since at this point it is single threaded)

hmwverbeek commented on src/org/processmining/plugins/etm/fitness/TraceGroupsCosts.java:
Do you need the trace name (traceName) here? As far as I can see, you create a map from the trace names to these objects, hence each trace name is mapped onto an object that contains it trace name.

msmuskan replied:
I will remove it, thanks for pointing out.

hmwverbeek commented on src/org/processmining/plugins/etm/fitness/metrics/FairnessReplay.java:
What I get from the code is that (a subset of) the traces are partitioned into two sets: Those where fairness:group=True (group A) and those where fairness:group is something else (group B). You then compare the average costs for both groups, and compute a fairness metric from these two averages. 

Wouldn't is be easier to define the second group as those traces that either do not have fairness:group set or where it does not have the value True when set? Then you would be sure that both sets partition the entire log.

As you are using the concept:name attribute to identify a trace, you require a log for which every trace has a unique concept:name. But it seems there is no good alternative available for this at the moment. 

I do not completely get the "1 + " on lines 418 and 419. I assume there is a good reason for adding 1 to the averages, but for me it is not clear.

msmuskan replied:
Actually, I think it is beneficial to require every trace to have a groupName set, because if its not then we can debug the event log and see why some traces are missing this information. So, we do not want to assume the missing groupName to be associated with the second group, because we cannot tell for sure. 

For now, I am using only binary groupName i.e. True or False, where True signifies sensitive or minority group. In the lines 330 to 333, you can see, I am first checking if groupName is True, and if not then I assume it is False. This seems similar to what you are suggesting. Note, later we might upgrade to multiple groups and not just the binary (true/false). 

The addition of 1 is to avoid the issue of division by zero. But now I am realizing I should just add an if condition to overcome this, because adding 1 to both averages can affect the calculation of the fairness, even though it might not be a significant change.

hmwverbeek commented on tests/bin-test/org/processmining/tests/newpackage/NewPackageTest.class:
This one should not be included in a commit, I guess.

hmwverbeek reviewed (commented):
See the comments for the files. 
Except for the class file, it would be okay to merge this, but there may be minor improvements possible.

msmuskan committed:
Added review changes

- Removed 1 from average costs.
- Removed traceName from the class.
- Removed the .class file.

msmuskan committed:
Added fairness based on distance, renamed old one

- Added fairness based on some relative distance, FaD = 1 - |(x-y)/(x+y+1)|
- Removed inclusion of fairness by default, can now be added from the drop down menu in ProM ETM plugin.

fmannhardt commented:
Merging it to have the code ready for ICPM. Future work is to integrate into the release.
